File: C/test/BasicTest1.c
#include "test.h"

#include "../logger/logger.h"

/**
 * Test basique 1 :
 * Initialise le logger, utilise tout les niveaux de logs et exit le logger.
 *
 * Conditions de réussite :
 * - Le dossier logs est créé.
 * - Il contient un seul fichier .log.
 * - Le fichier .log contient 8 lignes de logs.
 */
bool runBasicTest1() {
    LoggerType types[] = {INFO, SUCCESS, ERROR, WARNING, DEBUG};
    logger_init(FILE_AND_CONSOLE, types);

    INFO_LOG(FILE_AND_CONSOLE, "Info message");
    SUCCESS_LOG(FILE_AND_CONSOLE, "Success message");
    ERROR_LOG(FILE_AND_CONSOLE, "Error message");
    WARNING_LOG(FILE_AND_CONSOLE, "Warning message");
    DEBUG_LOG(FILE_AND_CONSOLE, "Debug message");

    logger_exit();

    // ====================

    // Le dossier logs est créé.
    struct stat buffer;
    if (stat("logs", &buffer) != 0 || !S_ISDIR(buffer.st_mode)) {
        return false;
    }

    // Il contient un seul fichier .log.
    DIR *dir = opendir("logs");
    if (dir == NULL) {
        return false;
    }
    struct dirent *ent;
    int nbFiles = 0;
    char *fileName;
    while ((ent = readdir(dir)) != NULL) {
        if (strcmp(ent->d_name, ".") != 0 && strcmp(ent->d_name, "..") != 0) {
            nbFiles++;
            fileName = ent->d_name;
        }
    }
    closedir(dir);
    if (nbFiles != 1) {
        return false;
    }

    // Le fichier .log contient 8 lignes de logs.
    char *filePath = malloc(sizeof(char) * (strlen("logs/") + strlen(fileName) + 1));
    sprintf(filePath, "logs/%s", fileName);
    FILE *file = fopen(filePath, "r");
    if (!file) {
        return false;
    }
    int nbLines = 0;
    char *line;
    size_t len = 0;
    while (getline(&line, &len, file) != -1) {
        nbLines++;
    }
    fclose(file);
    if (nbLines != 8) {
        return false;
    }

    return true;
}

Test BasicTest1 = {
        .name = "BasicTest1",
        .before = nothing,
        .run = runBasicTest1,
        .after = clearDir,
};

File: Go/test/BasicTest1.go
package main

import (
	"bufio"
	"io/ioutil"
	logger "logger"
	"os"
)

// BasicTest1 :
// Initialise le logger, utilise tout les niveaux de logs et exit le logger.
//
// Conditions de réussite :
// - Le dossier logs est créé.
// - Il contient un seul fichier .log.
// - Le fichier .log contient 8 lignes de logs.
var BasicTest1 = Test{
	Name: "BasicTest1",
	Before: func() {
	},
	Run: func() bool {
		logger.Init(logger.FILE_AND_CONSOLE, []logger.LoggerType{logger.DEBUG, logger.INFO, logger.WARNING, logger.ERROR, logger.SUCCESS})

		logger.Info("Info message")
		logger.Success("Success message")
		logger.Error("Error message")
		logger.Warning("Warning message")
		logger.Debug("Debug message")

		logger.Exit()

		// ====================

		// Le dossier logs est créé
		if _, err := os.Stat("logs"); os.IsNotExist(err) {
			return false
		}

		// Il contient un seul fichier .log.
		files, err := ioutil.ReadDir("logs")
		if err != nil {
			return false
		}
		if len(files) != 1 {
			return false
		}
		var fileName = files[0].Name()

		// Le fichier .log contient 8 lignes de logs.
		file, err := os.Open("logs/" + fileName)
		if err != nil {
			return false
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		var nbLine = 0
		for scanner.Scan() {
			nbLine++
		}
		if nbLine != 8 {
			return false
		}

		return true
	},
	After: func() {
		ClearDir("logs")
	},
}
